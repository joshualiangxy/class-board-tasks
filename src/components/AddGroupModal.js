import React, { useState } from 'react';
import Modal from 'react-modal';
import { startNewGroup } from '../actions/groups';
import { connect } from 'react-redux';

// TODO: Add the ability to add other users into the group
const AddGroupModal = ({ isOpen, onRequestClose, startNewGroup }) => {
  const [groupName, setGroupName] = useState('');
  const [module, setModule] = useState('');
  const [error, setError] = useState('');

  const onGroupNameChange = e => setGroupName(e.target.value);

  const onModuleChange = e => {
    const module = e.target.value.toUpperCase();

    setModule(module);
  };

  const onCancel = () => {
    setError('');
    setGroupName('');
    setModule('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedName = groupName.trim();
    const submittedMod = module.trim();
    if (!submittedName) {
      setError('Please enter a group name');
    } else {
      setError('');
      setGroupName('');
      setModule('');
      startNewGroup(submittedName, submittedMod).then(() => onRequestClose());
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={onCancel}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="Group name (required)"
          value={groupName}
          onChange={onGroupNameChange}
          autoFocus
        />
        <input
          type="text"
          placeholder="Module code/Class name (Optional)"
          value={module}
          onChange={onModuleChange}
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startNewGroup: (groupName, module) =>
    dispatch(startNewGroup(groupName, module))
});

export default connect(null, mapDispatchToProps)(AddGroupModal);
