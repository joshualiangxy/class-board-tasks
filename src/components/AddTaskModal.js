import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { startAddPersonalTask, startAddGroupTask } from '../actions/tasks';
import TaskForm from './TaskForm';

export const AddTaskModal = ({
  startAddPersonalTask,
  startAddGroupTask,
  removeAdmin,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  addGroupTask,
  users
}) => {
  const submitTask = task =>
    gid
      ? startAddGroupTask(task, groupName)
          .then(isAdmin => {
            if (isAdmin) return addGroupTask(task);
            else return removeAdmin();
          })
          .then(() => onRequestClose())
      : startAddPersonalTask(task).then(() => onRequestClose());

  return (
    <Modal
      className="task-modal"
      isOpen={isOpen}
      contentLabel="Add Task"
      onRequestClose={onRequestClose}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <h1>Add Task</h1>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startAddPersonalTask: task => dispatch(startAddPersonalTask(task)),
  startAddGroupTask: (task, groupName) =>
    dispatch(startAddGroupTask(task, groupName))
});

export default connect(undefined, mapDispatchToProps)(AddTaskModal);
