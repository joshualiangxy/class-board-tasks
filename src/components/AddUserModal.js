import React, { useState } from 'react';
import Modal from 'react-modal';
import firebase from '../firebase/firebase';

// Note: This still works off uid for now. This also does NOT work yet.
const AddUserModal = ({ isOpen, onRequestClose, group }) => {
  const addNewUser = (userEmail, group) => {
    const uid = userEmail; // TODO: Change this when email-uid are linked!

    // Both of these references are DocumentReference
    const userRef = firebase.firestore().collection('users').doc(uid);
    const groupRef = firebase.firestore().collection('groups').doc(group);

    // Gets the data from a user.
    userRef.get().then(doc => {
      const data = doc.data();
      // Write displayName, studentNum, and UID into a new document
      if (typeof data === 'undefined') {
        // Most likely a user that doesn't exist.
        setError('No such user found');
      } else {
        groupRef
          .collection('users')
          .doc(uid) // Create document for this user's UID
          .set({
            studentNum: data.studentNum,
            displayName: data.displayName,
            uid: uid,
            admin: false
          })
          .then(() => {
            setError('User added!');
          })
          .catch(error => {
            setError('Something went wrong.');
          });
      }
      // Defaults to non-admin.
    });
  };

  // TODO: When calling this modal, pass in the group UID as a prop
  const [userEmail, setUserEmail] = useState('');
  const [error, setError] = useState('');
  // const [usersAdded, setUsersAdded] = useState([]); // For listing down people that have been added?

  const onUserEmailChange = e => {
    setUserEmail(e.target.value);
  };

  const onCancel = () => {
    setUserEmail('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedEmail = userEmail.trim();
    if (!submittedEmail) {
      setError('Please enter a group name');
    } else {
      addNewUser(submittedEmail, group);
      // TODO: Catch user not found
      // Does not close after submitting, so that the user can continue to add more users

      // setUserEmail(''); // Also resets the field.
      // setError('Added!');
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={() => onRequestClose()}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="User ID (required)"
          value={userEmail}
          onChange={onUserEmailChange}
          autoFocus
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

export default AddUserModal;
