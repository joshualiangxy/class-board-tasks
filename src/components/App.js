import React from 'react';
import '../styles/App.css';

const App = () => (
  <div>
    <h1>This is my app</h1>
  </div>
);

export default App;

