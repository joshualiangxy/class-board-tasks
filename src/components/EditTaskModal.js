import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import {
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks
} from '../actions/tasks';
import TaskForm from './TaskForm';

export const EditTaskModal = ({
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  task,
  users,
  editGroupTask
}) => {
  const submitTask = updates =>
    gid
      ? startEditGroupTask(task.id, updates, groupName).then(() => {
          editGroupTask(task.id, updates);
          startSetTasks(); // Temporary solution
          onRequestClose();
        })
      : startEditPersonalTask(task.id, updates).then(() => onRequestClose());

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="Edit Task"
      onRequestClose={onRequestClose}
      appElement={document.getElementById('root')}
    >
      <h2>Edit Task</h2>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
        task={task}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startEditPersonalTask: (id, updates) =>
    dispatch(startEditPersonalTask(id, updates)),
  startEditGroupTask: (id, updates, groupName) =>
    dispatch(startEditGroupTask(id, updates, groupName)),
  startSetTasks: () => dispatch(startSetTasks())
});

export default connect(undefined, mapDispatchToProps)(EditTaskModal);
