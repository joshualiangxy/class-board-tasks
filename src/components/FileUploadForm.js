import React, { useState } from 'react';
import firebase from '../firebase/firebase';
import { startUpdateDownloadURL } from '../actions/tasks';
import { uploadFile, startRemoveUserFile } from '../actions/files';
import { connect } from 'react-redux';

export const FileUploadForm = ({
  id,
  uid,
  uploadFile,
  TaskEvent,
  startUpdateDownloadURL,
  startRemoveUserFile,
  fileExists
}) => {
  const [error, setError] = useState('');
  const [uploading, setUploading] = useState(false);

  const fileUpload = e => {
    e.preventDefault();
    e.persist();

    const element = document.getElementById(id);
    const files = element.files;
    if (files.length > 0) {
      const file = files[0];

      Promise.resolve()
        .then(() => {
          if (fileExists) return startRemoveUserFile(uid);
        })
        .then(() => {
          const uploadTask = uploadFile(file);

          uploadTask.on(
            TaskEvent.STATE_CHANGED,
            () => setUploading(true),
            error => {
              switch (error.code) {
                case 'storage/unauthorized':
                  setError('Not allowed to upload file!');
                  break;
                case 'storage/canceled':
                  setError('File upload cancelled');
                  break;
                default:
                  setError('File upload failed');
                  break;
              }
              setUploading(false);
            },
            () => {
              const ref = uploadTask.snapshot.ref;
              const fileName = ref.name;

              ref
                .getDownloadURL()
                .then(downloadURL => {
                  startUpdateDownloadURL(downloadURL, fileName);
                })
                .then(() => {
                  setError('');
                  setUploading(false);
                  e.target.reset();
                });
            }
          );
        });
    } else setError('No file selected');
  };

  return (
    <form onSubmit={fileUpload}>
      {error && <p>{error}</p>}
      <input
        type="file"
        id={id}
        onClick={e => e.stopPropagation()}
        accept=".pdf"
        disabled={uploading}
      />
      <button onClick={e => e.stopPropagation()} disabled={uploading}>
        {uploading ? 'Uploading...' : 'Upload'}
      </button>
    </form>
  );
};

const mapStateToProps = ({ tasks, auth }, { id }) => {
  const uid = auth.user.uid;
  const task = tasks.find(task => task.id === id);

  return {
    TaskEvent: firebase.storage.TaskEvent,
    fileExists: task ? task.downloadURLs.hasOwnProperty(uid) : false,
    uid
  };
};

const mapDispatchToProps = (dispatch, { id, gid, namingConvention }) => ({
  uploadFile: file => dispatch(uploadFile(file, id, namingConvention)),
  startUpdateDownloadURL: (downloadURL, fileName) =>
    dispatch(startUpdateDownloadURL(id, gid, downloadURL, fileName)),
  startRemoveUserFile: uid => dispatch(startRemoveUserFile(id, uid))
});

export default connect(mapStateToProps, mapDispatchToProps)(FileUploadForm);
