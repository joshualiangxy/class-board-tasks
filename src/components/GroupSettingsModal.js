import React, { useState } from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { addNewUser, getUser, recheckAdmin } from '../actions/groups';
import GroupUserListing from './GroupUserListing';

export const GroupSettingsModal = ({
  isOpen,
  onRequestClose,
  group,
  users,
  setUsers,
  admin,
  setAdmin,
  uid,
  kickUserLocal
}) => {
  const [userEmail, setUserEmail] = useState('');
  const [error, setError] = useState('');

  const onUserEmailChange = e => {
    setUserEmail(e.target.value);
  };

  const onCancel = () => {
    setUserEmail('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedEmail = userEmail.trim().toLowerCase();

    if (!submittedEmail) {
      setError('Please enter an email');
    } else {
      // Retrieve the user's data, and add it to the array of users.
      // user contains {admin, displayName, studentNum, uid} fields
      return recheckAdmin(uid, group).then(isAdmin => {
        if (isAdmin) {
          return getUser(submittedEmail).then(user => {            
            const newuser = {...user, admin: false}
            if (typeof user !== 'undefined') {
              return addNewUser(newuser, group)
                .then(() => {
                  setUsers([...users, newuser]);
                  setError('Added!');
                  setUserEmail('');
                })
                .catch(() => {
                  setError('Something went wrong');
                });
            } else {
              setError('No user found');
            }
          });
        } else {
          // This user is not an admin anymore
          setAdmin(false);
          setError('You are not an admin!');
        }
      });
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="Settings"
      onRequestClose={() => onRequestClose()}
      appElement={document.getElementById('root')}
      className="setting-modal"
      closeTimeoutMS={200}
    >
      <h1>Group Settings</h1>
      <div>
        {admin && (
          <form onSubmit={onSubmit} className="form">
            <input
              type="text"
              className="text-input"
              placeholder="Email (required)"
              value={userEmail}
              onChange={onUserEmailChange}
              autoFocus
            />
            {error && <p className="form__error">{error}</p>}
            <div>
              <button className="button button--norm">Add to list</button>{' '}
              <button onClick={onCancel} className="button button--norm">
                Cancel
              </button>
            </div>
          </form>
        )}
        {!admin && (
          <button onClick={onCancel} className="button button--norm">
            Close
          </button>
        )}
        <h2>Users:</h2>
        <GroupUserListing
          users={users}
          setUsers={setUsers}
          group={group}
          admin={admin}
          setAdmin={setAdmin}
          uid={uid}
          kickUserLocal={kickUserLocal}
          setError={setError}
        />
      </div>
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  addNewUser: (uid, gid) => dispatch(addNewUser(uid, gid))
});

export default connect(null, mapDispatchToProps)(GroupSettingsModal);
