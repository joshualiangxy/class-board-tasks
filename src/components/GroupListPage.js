import React, { useState } from 'react';
import AddGroupModal from './AddGroupModal';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export const GroupListPage = ({ groups }) => {
  const [open, setOpen] = useState(false);

  const onRequestClose = () => {
    setOpen(false);
  };

  const openNewGroup = () => {
    setOpen(true);
  };

  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">Groups</h1>
          <button
            className="button button--norm page-header__actions"
            onClick={openNewGroup}
          >
            Add New Group
          </button>
        </div>
      </div>
      <div className="content-container">
        {
          // Add a 'no groups' when there's no groups
          groups.length === 0 ? (
            <p className="list-header__text">No groups</p>
          ) : (
            groups.map(group => (
              <div key={group.gid} className="grouplist">
                <Link
                  className="button--link-black"
                  to={`/groups/${group.gid}`}
                >
                  <h2 className="groupitem">{group.name}</h2>
                  
                </Link>
              </div>
            ))
          )
        }
      </div>

      <AddGroupModal isOpen={open} onRequestClose={onRequestClose} />
    </div>
  );
};

const mapStateToProps = state => ({ groups: state.groups });

export default connect(mapStateToProps)(GroupListPage);
