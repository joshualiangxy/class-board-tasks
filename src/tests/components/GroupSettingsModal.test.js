import React from 'react';
import { shallow } from 'enzyme';
import { GroupSettingsModal } from '../../components/GroupSettingsModal';
import groupUsers from '../fixtures/groupUsers';

let wrapper;
const isOpen = true;
const onRequestClose = jest.fn();
const groupOne = 'one';
const groupTwo = 'two';
const users = groupUsers;
const setUsers = jest.fn(users => true);
const admin = true;
const setAdmin = jest.fn();
const kickUserLocal = jest.fn();
const uid1 = 'testuid';
const uid2 = 'testuid2';
const newUser = {
  displayName: 'Clarisse',
  studentNum: 'A0178221B',
  uid: 'testuid3'
};
const addNewUser = jest.fn((uid, gid) => Promise.resolve());
const getUser = jest.fn(email => {
  switch (email) {
    case 'email':
      return Promise.resolve(newUser);
    default:
      return Promise.resolve(undefined);
  }
});
const recheckAdmin = jest.fn((uid, gid) => {
  switch (uid) {
    case uid1:
      return Promise.resolve(true);
    case uid2:
      return Promise.resolve(false);
    default:
      return Promise.resolve(false);
  }
});

beforeEach(() => {
  jest.clearAllMocks();
  wrapper = shallow(
    <GroupSettingsModal
      isOpen={isOpen}
      onRequestClose={onRequestClose}
      group={groupOne}
      users={users}
      setUsers={setUsers}
      admin={admin}
      setAdmin={setAdmin}
      uid={uid1}
      kickUserLocal={kickUserLocal}
    />
  );
});

describe('render', () => {
  it('should render for admins', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input')).toHaveLength(1);
    expect(wrapper.find('input').at(0).prop('placeholder')).toEqual(
      'Email (required)'
    );
    expect(wrapper.find('Connect(GroupUserListing)')).toHaveLength(1);
  });

  it('should render for non-admins', () => {
    wrapper.setProps({ admin: false });
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input')).toHaveLength(0);
    expect(wrapper.find('Connect(GroupUserListing)')).toHaveLength(1);
  });
});

describe('addUser', () => {
  it('should change the input value when typed into', () => {
    expect(wrapper.find('input').at(0).prop('value')).toEqual('');
    wrapper.find('input').prop('onChange')({ target: { value: 'email' } });
    expect(wrapper.find('input').at(0).prop('value')).toEqual('email');
  });
  it('should set the error for invalid emails', () => {
    expect(wrapper.find('div.error').text()).toEqual('')
    wrapper.find('input').prop('onChange')({ target: { value: ' ' } });
    wrapper.find('form').at(0).prop('onSubmit')({ preventDefault: jest.fn() });
    expect(wrapper.find('div.error').text()).toEqual('Please enter an email')
  });

  // it('should set the error for invalid users (with valid email)', () => {
  //   expect(wrapper.find('div.error').text()).toEqual('')
  //   wrapper.find('input').prop('onChange')({ target: { value: 'validemail' } });
  //   wrapper.find('form').at(0).prop('onSubmit')({ preventDefault: jest.fn() });
  //   expect(recheckAdmin).toHaveBeenCalledTimes(1)
  //   expect(wrapper.find('div.error').text()).toEqual('No user found')
  
  // })
});
