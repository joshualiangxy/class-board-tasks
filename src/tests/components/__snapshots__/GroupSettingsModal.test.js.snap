// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render for admins 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="New Group"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div>
    <form
      onSubmit={[Function]}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Email (required)"
        type="text"
        value=""
      />
      <button>
        Add to list
      </button>
      <div
        className="error"
      />
    </form>
    <button
      onClick={[Function]}
    >
      Close
    </button>
    <h3>
      Users:
    </h3>
    <Connect(GroupUserListing)
      admin={true}
      group="one"
      kickUserLocal={[MockFunction]}
      setAdmin={[MockFunction]}
      setError={[Function]}
      setUsers={[MockFunction]}
      uid="testuid"
      users={
        Array [
          Object {
            "admin": true,
            "displayName": "Robert",
            "studentNum": "A0123456A",
            "uid": "testuid",
          },
          Object {
            "admin": false,
            "displayName": "Peter",
            "studentNum": "A0000000A",
            "uid": "testuid2",
          },
        ]
      }
    />
  </div>
</Modal>
`;

exports[`render should render for non-admins 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="New Group"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div>
    <button
      onClick={[Function]}
    >
      Close
    </button>
    <h3>
      Users:
    </h3>
    <Connect(GroupUserListing)
      admin={false}
      group="one"
      kickUserLocal={[MockFunction]}
      setAdmin={[MockFunction]}
      setError={[Function]}
      setUsers={[MockFunction]}
      uid="testuid"
      users={
        Array [
          Object {
            "admin": true,
            "displayName": "Robert",
            "studentNum": "A0123456A",
            "uid": "testuid",
          },
          Object {
            "admin": false,
            "displayName": "Peter",
            "studentNum": "A0000000A",
            "uid": "testuid2",
          },
        ]
      }
    />
  </div>
</Modal>
`;
